src/ynEIGEN/AssetRegistry.sol

```diff
@@ -0,0 +1,358 @@
+    //--------------------------------------------------------------------------------------
+
+     /// @notice List of supported ERC20 asset contracts.
+    IERC20[] public assets;
+
+    mapping(IERC20 => AssetData) public _assetData;
+
+    bool public actionsPaused;
+
+    IRateProvider public rateProvider;
+    IEigenStrategyManager public eigenStrategyManager;
+    IynEigen ynEigen;
+
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  INITIALIZATION  ----------------------------------
+    //--------------------------------------------------------------------------------------
+
+    constructor() {
+       _disableInitializers();
+    }
+
+    struct Init {
+        IERC20[] assets;
+        IRateProvider rateProvider;
+        IEigenStrategyManager eigenStrategyManager;
+        IynEigen ynEigen;
+        address admin;
+        address pauser;
+        address unpauser;
+        address assetManagerRole;
+    }
+
+    function initialize(Init calldata init)
+        public
+        notZeroAddress(address(init.rateProvider))
+        notZeroAddress(address(init.admin))
+        notZeroAddress(address(init.pauser))
+        notZeroAddress(address(init.unpauser))
+        initializer {
+        __AccessControl_init();
+        __ReentrancyGuard_init();
+
+        _grantRole(DEFAULT_ADMIN_ROLE, init.admin);
+        _grantRole(PAUSER_ROLE, init.pauser);
+        _grantRole(UNPAUSER_ROLE, init.unpauser);
+        _grantRole(ASSET_MANAGER_ROLE, init.assetManagerRole);
+        
+        uint256 assetsLength = init.assets.length;
+        for (uint256 i = 0; i < assetsLength; i++) {
+            if (address(init.assets[i]) == address(0)) {
+                revert ZeroAddress();
+            }
+            if (_assetData[init.assets[i]].active) {
+                revert AssetAlreadyActive(address(init.assets[i]));
+            }
+            assets.push(init.assets[i]);
+            _assetData[init.assets[i]] = AssetData({
+                active: true
+            });
+        }
+
+        rateProvider = init.rateProvider;
+        eigenStrategyManager = init.eigenStrategyManager;
+        ynEigen = init.ynEigen;
+    }
+
+    //--------------------------------------------------------------------------------------
+    //----------------------------------  ASSET MANAGEMENT  --------------------------------
+    //--------------------------------------------------------------------------------------
+
+    /**
+     * @notice Adds a new asset to the system.
+     * @dev Adds an asset to the _assetData mapping and sets it as active. This function can only be called by the strategy manager.
+     * @param asset The address of the ERC20 token to be added.
+     */
+    function addAsset(IERC20 asset)
+    public
+    onlyRole(ASSET_MANAGER_ROLE)
+    notZeroAddress(address(asset))
+    whenNotPaused {
+        if (_assetData[asset].active) {
```

Comment:
Adding a disabled asset is possible, but it will be pushed to the array a second time. I suggest using [enums](https://solidity-by-example.org/enum/) instead of booleans to differentiate between non-existent assets and disabled assets.

Reply:
fixed, great catch, thanks